### DISCUSSION TITLE
Auth: resetPassword invalid token response status code from 500 to 400 or 403?

### DISCUSSION THREAD

**Discussion by edtorba** [payloadcms/payload#4909] [Feature Requests & Ideas]
When resetting the password via `resetPassword`, if an incorrect token is provided, an error is received from the API. While receiving a 500 status code is acceptable (ish?), it seems more appropriate in this instance to use a 400 (Bad Request) or 403 (Forbidden) status code.

The definition for the 500 status code is as follows:

> The HyperText Transfer Protocol (HTTP) 500 Internal Server Error server error response code indicates that the server encountered an unexpected condition that prevented it from fulfilling the request.

However, in our case, the server is aware of the situation, and I believe a 400 (Bad Request) status code, indicating that the server cannot or will not process the request due to something perceived as a client error, or a 403 (Forbidden) status code, indicating that the server understands the request but refuses to authorise it, would be more appropriate in this instance.

In terms of effort, the required change is straightforward; it simply involves adding an extra argument to set the appropriate status code.

resetPassword.ts

```ts
// ...

 const user = await payload.db.findOne<any>({
      collection: collectionConfig.slug,
      req,
      where: {
        resetPasswordExpiration: { greater_than: new Date() },
        resetPasswordToken: { equals: data.token },
      },
    })

    if (!user) throw new APIError('Token is either invalid or has expired.')

// ...
```
https://github.com/payloadcms/payload/blob/b40e9f85a299789dfb35016f4ab4f6a9e97698b6/packages/payload/src/auth/operations/resetPassword.ts#L70

APIError.ts
```ts
class APIError<
  TData extends null | object = { [key: string]: unknown } | null,
> extends ExtendableError<TData> {
  /**
   * Creates an API error.
   * @param {string} message - Error message.
   * @param {number} status - HTTP status code of error.
   * @param {object} data - response data to be returned.
   * @param {boolean} isPublic - Whether the message should be visible to user or not.
   */
  constructor(
    message: string,
    status: number = httpStatus.INTERNAL_SERVER_ERROR,
    data: TData = null,
    isPublic = false,
  ) {
    super(message, status, data, isPublic)
  }
}
```
https://github.com/payloadcms/payload/blob/b40e9f85a299789dfb35016f4ab4f6a9e97698b6/packages/payload/src/errors/APIError.ts#L41

**Comment by DanRibbens** [comment]
I agree with returning 403 status instead of 500. Because it is a breaking change though, I would open this as a PR against what will be the 3.0 branch when that is ready to contribute to.

Are you interested in making a PR?

**Reply by edtorba** [reply]
Sure, I will submit a pull request ~~later today~~ when  3.0 branch be available. 👍

**Reply by edtorba** [reply]
I'll review the other authentication-related actions as well 👍

**Reply by DanRibbens** [reply]
Beautiful, thanks! We're hoping to have an alpha of 3.0 sometime in February last I heard.

**Reply by edtorba** [reply]
@DanRibbens do you think the next-poc branch is in a good enough state for me to look into the status codes?

**Reply by nunoci** [reply]
Any updates on this subject?
We running payload version 2.11.1 instance  and it keep crashing on railway.

**Reply by edtorba** [reply]
@nunoci that was updated, but only in v3 https://github.com/payloadcms/payload/pull/5254

**Reply by nunoci** [reply]
@edtorba thank you for your answer...
Do you have any plan to patch on version 2? 
Because the solution for current release since 3.0 is beta and we have clients in production now (version 2x stable).
railway keeps crashing when resetPassword have an invalid token.

**Reply by edtorba** [reply]
I personally wouldn’t have a problem creating a PR for version 2, but it’s more a question for @DanRibbens to see if they would be open to including this change in version 2, as the initial response suggested doing it only in version 3.

**Reply by edtorba** [reply]
If Dan gives me the thumbs up, I can have a PR ready this evening.👍

**Reply by DanRibbens** [reply]
It's hard to say if this is a breaking change or a bug fix. I'm leaning towards leaving it as is.

**Reply by edtorba** [reply]
I think it’s actually a bit of both, a breaking change and a bug fix.

**Reply by nunoci** [reply]
How to avoid the server stop responding when the auth fails? ** on login, or on reset pass**  I have to restart the server 4 times a day...
![Screenshot 2024-08-05 at 5 09 43 PM](https://github.com/user-attachments/assets/e972a0cb-391d-45ee-aa5a-8a4dea6fa439)

**Reply by edtorba** [reply]
@nunoci I'd try setting up a custom global error handler in your Express server configuration (server.ts file).

```js
const app = express();

// ... your stuff.

// Error handling middleware
app.use((err, req, res, next) => {
  console.error(err.stack); // Log the error stack
  res.status(500).send('Internal Server Error'); // Send a generic response
});

// "app..listen(..." stuff
```

**Reply by edtorba** [reply]
@nunoci did that work? 👀

**Reply by nuno** [reply]
> @nunoci did that work? 👀

It is just to paste that code before `app.listen(3001);` that is it?

**Reply by nunoci** [reply]
Does not work at all.
